server/entity/firework.go

```diff
@@ -0,0 +1,179 @@
+package entity
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/block/cube/trace"
+	"github.com/df-mc/dragonfly/server/entity/damage"
+	"github.com/df-mc/dragonfly/server/internal/nbtconv"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/sound"
+	"github.com/go-gl/mathgl/mgl64"
+	"math"
+	"math/rand"
+)
+
+// Firework is an item (and entity) used for creating decorative explosions, boosting when flying with elytra, and
+// loading into a crossbow as ammunition.
+type Firework struct {
+	transform
+
+	yaw, pitch float64
+	firework   item.Firework
+
+	owner world.Entity
+
+	c *MovementComputer
+
+	ticks int
+	close bool
+}
+
+// NewFirework ...
+func NewFirework(pos mgl64.Vec3, yaw, pitch float64, firework item.Firework) *Firework {
+	f := &Firework{
+		yaw:      yaw,
+		pitch:    pitch,
+		firework: firework,
+		c:        &MovementComputer{},
+		ticks:    int(firework.RandomizedDuration().Milliseconds() / 50),
+	}
+	f.transform = newTransform(f, pos)
+	f.vel = mgl64.Vec3{rand.Float64() * 0.001, 0.05, rand.Float64() * 0.001}
+	return f
+}
+
+// Name ...
+func (f *Firework) Name() string {
+	return "Firework Rocket"
+}
+
+// EncodeEntity ...
+func (f *Firework) EncodeEntity() string {
+	return "minecraft:fireworks_rocket"
+}
+
+// BBox ...
+func (f *Firework) BBox() cube.BBox {
+	return cube.BBox{}
+}
+
+// Firework returns the underlying item.Firework of the Firework.
+func (f *Firework) Firework() item.Firework {
+	return f.firework
+}
+
+// Rotation ...
+func (f *Firework) Rotation() (float64, float64) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	return f.yaw, f.pitch
+}
+
+// Tick ...
+func (f *Firework) Tick(w *world.World, current int64) {
+	if f.close {
+		_ = f.Close()
+		return
+	}
+
+	f.mu.Lock()
+	f.vel[0] *= 1.15
+	f.vel[1] += 0.04
+	f.vel[2] *= 1.15
+	m := f.c.TickMovement(f, f.pos, f.vel, f.yaw, f.pitch)
+	f.pos, f.vel, f.yaw, f.pitch = m.pos, m.vel, m.yaw, m.pitch
+	f.mu.Unlock()
+
+	m.Send()
+
+	if m.pos[1] < float64(w.Range()[0]) && current%10 == 0 {
+		f.close = true
+		return
+	}
+
+	f.ticks--
+	if f.ticks < 0 {
```

Comment:
Would invert this check and return to reduce indentation.

Reply:
Inverted.